File: python/bellman_ford.py
def bellman_ford_path(edges, start):
    nodes = set()
    for u, v, w in edges:
        nodes.add(u)
        nodes.add(v)

    dist = {n: float('inf') for n in nodes}
    dist[start] = 0
    V = len(nodes)

    for _ in range(V - 1):
        for u, v, w in edges:
            if dist[u] != float('inf') and dist[u] + w < dist[v]:
                dist[v] = dist[u] + w

    for u, v, w in edges:
        if dist[u] != float('inf') and dist[u] + w < dist[v]:
            return "Negative Cycle Detected"

    return dist

# Example
E = [
    ('A', 'B', -1),
    ('A', 'C', 4),
    ('B', 'C', 3),
    ('B', 'D', 2),
    ('B', 'E', 2),
    ('D', 'B', 1),
    ('E', 'D', -3)
]

start_node = 'A'
res = bellman_ford_path(E, start_node)
print(f"Shortest Paths from {start_node}: {res}")

File: php/bellman_ford.php
<?php

function bellmanFord($vertices, $edges, $src) {
    // Step 1: Initialize distances
    $dist = array_fill(0, $vertices, INF);
    $dist[$src] = 0;

    // Step 2: Relax edges |V| - 1 times
    for ($i = 0; $i < $vertices - 1; $i++) {
        $updated = false;
        foreach ($edges as [$u, $v, $w]) {
            if ($dist[$u] != INF && $dist[$u] + $w < $dist[$v]) {
                $dist[$v] = $dist[$u] + $w;
                $updated = true;
            }
        }
        if (!$updated) break; // Optimization: stop if no change
    }

    // Step 3: Check for negative-weight cycles
    foreach ($edges as [$u, $v, $w]) {
        if ($dist[$u] != INF && $dist[$u] + $w < $dist[$v]) {
            echo "Graph contains a negative weight cycle!\n";
            return null;
        }
    }

    return $dist;
}

// Example graph
$vertices = 5;
$edges = [
    [0, 1, -1],
    [0, 2, 4],
    [1, 2, 3],
    [1, 3, 2],
    [1, 4, 2],
    [3, 2, 5],
    [3, 1, 1],
    [4, 3, -3]
];

$src = 0;
$distances = bellmanFord($vertices, $edges, $src);

if ($distances !== null) {
    echo "Vertex\tDistance from Source ($src)\n";
    foreach ($distances as $i => $d) {
        $dStr = ($d == INF) ? "∞" : $d;
        echo "$i\t$dStr\n";
    }
}
